src/Lecture4.hs

```diff
@@ -135,7 +140,33 @@ errors. We will simply return an optional result here.
 -}
 
 parseRow :: String -> Maybe Row
-parseRow = error "TODO"
+parseRow = readRow . splitCSVLine
+   where readRow :: [String] -> Maybe Row
+         readRow cols
+            | length cols /= 3 = Nothing
+            | otherwise = Row <$> readProduct c1 <*> readTradeType c2 <*> readInt c3
+               where c1 = head cols
+                     c2 = cols !! 1
+                     c3 = last cols
```

Comment:
You can replace `length` and usage of `head`, `last` and `!!` with a single pattern on the list on the top level. So instead of checking the value separately and then extracting list parts using unsafe functions, you can pattern match on the desired shape:

```suggestion
         readRow cols = case cols of
             [c1, c2, c3] -> Row <$> readProduct c1 <*> readTradeType c2 <*> readInt c3
             _            -> Nothing
```

Reply:
nice!!! I knew I could avoid the head and last somehow , pattern matching !